docs/design/metainfo-rpc.md

```diff
@@ -26,87 +26,211 @@ The following is a proto file for just the rpc portion to create segments. The r
 With this design, streams will be created that contain multiple parts. There were already RPCs to finalize individual segments, but each segment was not logically associated with a part, but instead just the specific object.
 
 ```protobuf
-message AddressedOrderLimit {
-    orders.OrderLimit2 limit = 1;
-    node.NodeAddress storage_node_address = 2;
+service Metainfo {
+    rpc CreateBucket(BucketCreateRequest) returns (BucketCreateResponse);
+    rpc GetBucket(BucketGetRequest) returns (BucketGetResponse);
+    rpc DeleteBucket(BucketDeleteRequest) returns (BucketDeleteResponse);
+    rpc ListBuckets(BucketListRequest) returns (BucketListResponse);
+
+    rpc CreateObject(ObjectCreateRequest) returns (ObjectCreateResponse);
+    rpc CommitObject(ObjectCommitRequest) returns (ObjectCommitResponse);
+    rpc ListObjects(ObjectListRequest) returns (ObjectListResponse);
+
+    rpc DeleteObject(ObjectDeleteRequest) returns (ObjectDeleteResponse);
+    rpc FinishDeleteObject(ObjectDeleteRequest) returns (ObjectDeleteResponse);
+
+    rpc CreateSegment(SegmentCreateRequest) returns (SegmentCreateResponse);
+    rpc CommitSegment(SegmentCommitRequest) returns (SegmentCommitResponse);
+    rpc ContinueSegment(SegmentContinueRequest) returns (SegmentContinueResponse);
```

Comment:
I think this is not needed and we can do an async create/commit in sequence instead.

Reply:
Since there aren't dependencies between segments, yes, there's no difference. It's probably easier and more bug free to just support the combined RPC, though, as you don't have to worry about partial failures on the client. For example, what if the start succeeds but the commit fails? Does one then need to cancel the start? What if the cancel fails? I'm sure answers can exist for those questions, but it'd be nice if they didn't need to be asked.

Reply:
The server needs to handle the partial failure mode, yes, but the client doesn't. From the client's perspective, a continue RPC could be atomic.

Reply:
If uplink crashes and wants to continue the upload it's pretty much the same state there will be multiple segments per stream that may have not finished uploading.